File: Python3/Max Number of K-Sum Pairs.py
#1
class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        cnt, ans = Counter(nums), 0
        for val in cnt:
            ans += min(cnt[val], cnt[k - val])
        return ans//2

#
2class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        nums.sort()
        l = 0
        r = len(nums)-1
        ans = 0
        while l<r:
            if nums[r]+nums[l]==k:
                ans+=1
                l+=1
                r-=1
            elif nums[r]+nums[l]>k:
                r-=1
            else:
                l+=1
                
                
        return ans

File: Golang/Max Number of K-Sum Pairs.go
//1
func maxOperations(nums []int, k int) int {
    sort.Ints(nums)
    left, right := 0, len(nums)-1

    count := 0
    for left < right {
        sum := nums[left] + nums[right]
        if sum == k {
            count++
            left++
            right--
        } else if sum < k {
            left++
        } else {
            right--
        }
    }

    return count
}

//2
func maxOperations(nums []int, k int) int {
  count := 0
	m := make(map[int]int)
	for _, num := range nums {
		if v, ok := m[k-num]; ok {
			if v > 0 {
				m[k-num]--
				m[num]--
				count++
			}
		}
		m[num]++
	}
	return count
    
}